tree/treeplayer/src/TCsvDS.cxx

```diff
@@ -0,0 +1,255 @@
+   }
+}
+
+void TCsvDS::FillRecord(std::string &line, Record &record)
+{
+   std::istringstream lineStream(line);
+   auto i = 0U;
+
+   auto columns = ParseColumns(line);
+
+   for (auto &col : columns) {
+      auto &colType = fColTypes[fHeaders[i]];
+
+      if (colType == "int") {
+         record.emplace_back(new int(std::stoi(col))); // TODO: LONG?
+      } else if (colType == "double") {
+         record.emplace_back(new double(std::stod(col)));
+      } else if (colType == "bool") {
+         bool *b = new bool();
+         record.emplace_back(b);
+         std::istringstream is(col);
+         is >> std::boolalpha >> *b;
+      } else {
+         record.emplace_back(new std::string(col));
+      }
+      ++i;
+   }
+}
+
+void TCsvDS::GenerateHeaders(size_t size)
+{
+   for (size_t i = 0; i < size; ++i) {
+      fHeaders.push_back("Col" + std::to_string(i));
+   }
+}
+
+std::vector<void *> TCsvDS::GetColumnReadersImpl(std::string_view colName, const std::type_info &)
+{
+   const auto &colNames = GetColumnNames();
+   const auto index = std::distance(colNames.begin(), std::find(colNames.begin(), colNames.end(), colName));
+   std::vector<void *> ret(fNSlots);
+   for (auto slot : ROOT::TSeqU(fNSlots)) {
+      ret[slot] = (void *)&fColAddresses[index][slot];
+   }
+   return ret;
+}
+
+void TCsvDS::InferColTypes(std::vector<std::string> &columns)
+{
+   auto i = 0U;
+   for (auto &col : columns) {
+      InferType(col, i);
+      ++i;
+   }
+}
+
+void TCsvDS::InferType(std::string &col, unsigned int idxCol)
+{
+   std::string type;
+   int dummy;
+
+   if (intRegex.Index(col, &dummy) != -1) {
+      type = "int";
+   } else if (doubleRegex1.Index(col, &dummy) != -1 || doubleRegex2.Index(col, &dummy) != -1) {
+      type = "double";
+   } else if (trueRegex.Index(col, &dummy) != -1 || falseRegex.Index(col, &dummy) != -1) {
+      type = "bool";
+   } else { // everything else is a string
+      type = "std::string";
+   }
+   // TODO: Date
+
+   fColTypes[fHeaders[idxCol]] = type;
+}
+
+std::vector<std::string> TCsvDS::ParseColumns(std::string &line)
+{
+   std::vector<std::string> columns;
+
+   for (size_t i = 0; i < line.size(); ++i) {
+      i = ParseValue(line, columns, i);
+   }
+
+   return columns;
+}
+
+size_t TCsvDS::ParseValue(std::string &line, std::vector<std::string> &columns, size_t i)
+{
+   std::stringstream val;
+   bool quoted = false;
+
+   for (; i < line.size(); ++i) {
+      if (line[i] == fDelimiter && !quoted) {
+         break;
+      } else if (line[i] == '"') {
+         // Keep just one quote for escaped quotes, none for the normal quotes
+         if (line[i + 1] != '"') {
+            quoted = !quoted;
+         } else {
+            val << line[++i];
+         }
+      } else {
+         val << line[i];
+      }
+   }
+
+   columns.emplace_back(val.str());
+
+   return i;
+}
+
+TCsvDS::TCsvDS(std::string_view fileName, bool readHeaders, char delimiter) // TODO: Let users specify types?
```

Comment:
This can be a variadic template constructor. We know how to expand packs de facto statically looping on types: https://github.com/root-project/root/blob/master/tree/treeplayer/inc/ROOT/TDFInterface.hxx#L210

Reply:
Since everything is at runtime anyway we could also just take a string like `"double:bool:string:double:double"`.

Reply:
That is a good point. This would allow to read whatever csv (read only mainly) adding on top headers.